File: sdks/py_sdk/sw4rm/cancellation.py
"""SW4-004 cancellation helpers for gateway implementations."""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Callable, Iterable, Optional

from sw4rm.protos import common_pb2, handoff_pb2


MIN_GRACE_PERIOD_MS = 5000


@dataclass
class CancellationFlag:
    """Cancellation state for one delegation correlation."""

    cancelled: bool
    grace_period_ms: int
    cancel_time_ms: int


class CancellationManager:
    """Tracks cancellation state and grace-expiry preemption decisions."""

    def __init__(self, *, time_ms_fn: Optional[Callable[[], int]] = None) -> None:
        self._time_ms = time_ms_fn or (lambda: int(time.time() * 1000))
        self.child_delegations: dict[str, list[str]] = {}
        self.cancellation_flags: dict[str, CancellationFlag] = {}

    def register_child_delegation(
        self, parent_correlation_id: str, child_correlation_id: str
    ) -> None:
        """Link parent/child delegation ids for cancellation cascade."""
        self.child_delegations.setdefault(parent_correlation_id, []).append(
            child_correlation_id
        )

    def handle_cancel_delegation(
        self, cancel: handoff_pb2.CancelDelegation
    ) -> handoff_pb2.CancelDelegationResponse:
        """Set cancellation flags for correlation and any known children."""
        grace_period_ms = max(int(cancel.grace_period_ms), MIN_GRACE_PERIOD_MS)
        cancel_time_ms = self._time_ms()

        self.cancellation_flags[cancel.correlation_id] = CancellationFlag(
            cancelled=True,
            grace_period_ms=grace_period_ms,
            cancel_time_ms=cancel_time_ms,
        )

        for child_corr in self.child_delegations.get(cancel.correlation_id, []):
            self.cancellation_flags[child_corr] = CancellationFlag(
                cancelled=True,
                grace_period_ms=grace_period_ms,
                cancel_time_ms=cancel_time_ms,
            )

        return handoff_pb2.CancelDelegationResponse(acknowledged=True)

    def is_cancelled(self, correlation_id: str) -> bool:
        """Return True when a cancellation flag exists and is active."""
        entry = self.cancellation_flags.get(correlation_id)
        return bool(entry and entry.cancelled)

    def is_grace_expired(
        self, correlation_id: str, *, now_ms: Optional[int] = None
    ) -> bool:
        """Return True when cancellation grace period has elapsed."""
        entry = self.cancellation_flags.get(correlation_id)
        if entry is None or not entry.cancelled:
            return False
        current_ms = self._time_ms() if now_ms is None else now_ms
        return (current_ms - entry.cancel_time_ms) >= entry.grace_period_ms

    def forced_preemption_error_code(
        self, correlation_id: str, *, now_ms: Optional[int] = None
    ) -> int:
        """Return FORCED_PREEMPTION once grace expires, otherwise unspecified."""
        if self.is_grace_expired(correlation_id, now_ms=now_ms):
            return common_pb2.FORCED_PREEMPTION
        return common_pb2.ERROR_CODE_UNSPECIFIED

    def collect_forced_preemptions(
        self, correlation_ids: Iterable[str], *, now_ms: Optional[int] = None
    ) -> set[str]:
        """Return cancelled correlations whose grace period has expired."""
        current_ms = self._time_ms() if now_ms is None else now_ms
        forced: set[str] = set()
        for correlation_id in correlation_ids:
            if self.is_grace_expired(correlation_id, now_ms=current_ms):
                forced.add(correlation_id)
        return forced

File: sdks/ex_sdk/lib/sw4rm/cancellation.ex
defmodule Sw4rm.Cancellation do
  @moduledoc """
  Standalone functional cancellation manager for SW4-004 conformance.

  Pure functional module with immutable state — conformance vectors create
  fresh state per test, so a GenServer is unnecessary here.
  """

  @min_grace_period_ms 5000

  defmodule CancellationFlag do
    @moduledoc "A single cancellation flag with timing information."
    defstruct [:correlation_id, :cancel_time_ms, :grace_period_ms, :reason]
  end

  @type state :: %{
          flags: %{String.t() => CancellationFlag.t()},
          children: %{String.t() => [String.t()]},
          now_ms_fn: (-> integer())
        }

  @doc "Create a new cancellation state."
  @spec new(keyword()) :: state()
  def new(opts \\ []) do
    %{
      flags: %{},
      children: %{},
      now_ms_fn: Keyword.get(opts, :now_ms_fn, fn -> System.system_time(:millisecond) end)
    }
  end

  @doc "Register a child delegation under a parent correlation ID."
  @spec register_child_delegation(state(), String.t(), String.t()) :: state()
  def register_child_delegation(state, parent, child) do
    children = Map.update(state.children, parent, [child], &[child | &1])
    %{state | children: children}
  end

  @doc "Handle a cancel delegation request. Returns `{:ok, updated_state}`."
  @spec handle_cancel_delegation(state(), map()) :: {:ok, state()}
  def handle_cancel_delegation(state, request) do
    correlation_id = request.correlation_id
    reason = Map.get(request, :reason, "")
    requested_grace = Map.get(request, :grace_period_ms, 0)
    effective_grace = max(requested_grace, @min_grace_period_ms)
    now_ms = state.now_ms_fn.()

    all_ids = collect_tree(correlation_id, state.children)
    direct_children = Map.get(state.children, correlation_id, [])

    flags =
      Enum.reduce(all_ids, state.flags, fn id, acc ->
        # Children get grace clamped to parent's remaining grace
        child_grace =
          if id in direct_children do
            # If the parent was already cancelled, clamp child grace to parent's remaining
            case Map.get(acc, correlation_id) do
              %CancellationFlag{cancel_time_ms: parent_cancel, grace_period_ms: parent_grace} ->
                parent_remaining = max(0, parent_cancel + parent_grace - now_ms)
                min(effective_grace, parent_remaining)

              nil ->
                effective_grace
            end
          else
            effective_grace
          end

        flag = %CancellationFlag{
          correlation_id: id,
          cancel_time_ms: now_ms,
          grace_period_ms: child_grace,
          reason: reason
        }

        Map.put(acc, id, flag)
      end)

    {:ok, %{state | flags: flags}}
  end

  @doc "Check if a correlation ID has been cancelled."
  @spec is_cancelled?(state(), String.t()) :: boolean()
  def is_cancelled?(state, correlation_id) do
    Map.has_key?(state.flags, correlation_id)
  end

  @doc "Check if the grace period has expired for a cancelled correlation."
  @spec is_grace_expired?(state(), String.t(), integer() | nil) :: boolean()
  def is_grace_expired?(state, correlation_id, now_ms \\ nil) do
    case Map.get(state.flags, correlation_id) do
      nil ->
        false

      flag ->
        check_time = now_ms || state.now_ms_fn.()
        check_time >= flag.cancel_time_ms + flag.grace_period_ms
    end
  end

  @doc "Get the forced preemption error code for a correlation."
  @spec forced_preemption_error_code(state(), String.t(), integer() | nil) :: integer()
  def forced_preemption_error_code(state, correlation_id, now_ms \\ nil) do
    if is_grace_expired?(state, correlation_id, now_ms) do
      Sw4rm.ErrorCodes.forced_preemption()
    else
      Sw4rm.ErrorCodes.error_code_unspecified()
    end
  end

  @doc "Collect correlation IDs whose grace period has expired from a list of active correlations."
  @spec collect_forced_preemptions(state(), [String.t()], integer() | nil) :: [String.t()]
  def collect_forced_preemptions(state, active_correlations, now_ms \\ nil) do
    Enum.filter(active_correlations, fn corr_id ->
      is_cancelled?(state, corr_id) && is_grace_expired?(state, corr_id, now_ms)
    end)
  end

  @doc "Get the effective grace period for a cancelled correlation."
  @spec effective_grace_period_ms(state(), String.t()) :: integer() | nil
  def effective_grace_period_ms(state, correlation_id) do
    case Map.get(state.flags, correlation_id) do
      nil -> nil
      flag -> flag.grace_period_ms
    end
  end

  @doc "Return the minimum grace period constant."
  @spec min_grace_period_ms() :: non_neg_integer()
  def min_grace_period_ms, do: @min_grace_period_ms

  # -- Helpers --

  defp collect_tree(id, children_map) do
    direct = Map.get(children_map, id, [])
    nested = Enum.flat_map(direct, &collect_tree(&1, children_map))
    [id | direct ++ nested] |> Enum.uniq()
  end
end
